doc/reference/trd-syscalls.md

```diff
@@ -309,12 +309,23 @@ memory and is not the Null Callback described below), the kernel MUST
 NOT invoke the requested driver and MUST immediately return a failure
 with a return code of EINVAL.
 
-A passed callback MUST be valid until the next invocation of `subscribe`
-with the same syscall and driver identifier. When userspace invokes
-subscribe, the kernel MUST cancel all pending callbacks for that driver
-and subscribe identifier: it MUST NOT invoke the previous callback after
-the call to subscribe, and MUST NOT invoke the new callback for events
-that occurred before the call to subscribe. 
+When the kernel invokes a driver, it passes a handle representing the passed
+callback to the driver. The driver is then responsible for storing the handle,
+and MUST return a handle, which is returned to the caller of `subscribe`. With
+the exception of a handle to the Null Callback, capsules cannot create or
+duplicate these handles.
+
+A passed callback MUST remain valid until all handles to that callback have been
+returned to userspace. When userspace invokes subscribe, the kernel MUST cancel
+all pending callbacks for that driver and subscribe identifier: it MUST NOT
+invoke the returned callback after the call to subscribe.
+
+Drivers SHOULD implement swapping semantics for the subscribe handles for each
```

Comment:
This should be a MUST, right? We can't statically check it, any more than we can statically check passed callbacks being valid. But we can state what the expectations are. A SHOULD means that it's OK if an implementation doesn't, as long as they really think they know what they're doing.

Reply:
Yes, if we're following the precise definitions of RFC 2119 a "SHOULD" is a too weak statement here and "MUST" seems more appropriate.